### DISCUSSION TITLE
HeatmapAnnotation not working within mapply call

### DISCUSSION THREAD

**Discussion by flde** [jokergoo/ComplexHeatmap#802] [Q&A] [answered]
Hi, 

Many thanks for your great work and tutorials. However, I encountered a problem I would need some help with. 

When I call `HeatmapAnnotation(foo1 = runif(10))` in my script everything works perfectly. But when I call it embedded in a function with mapply it breaks. 
`plot_FUN <- function(x, y) {HeatmapAnnotation(foo1 = runif(10))}`
`mapply(plot_FUN, list(a=1), list(a=1), SIMPLIFY = FALSE)`

> Error: It seems you are putting only one argument to the function. If it is a simple vector annotation or a function annotation (e.g. anno_*()), specify it as HeatmapAnnotation(name = value). If it is a data frame annotation, specify it as HeatmapAnnotation(df = value)

I thought it is a mapply problem but for example ComplexHeatmap::pheatmap works with mapply. And it works with lapply. Any ideas?

`lapply(list(a=1), plot_FUN)`

I work on JupyterLab with R 4.1.0 kernel and ComplexHeatmap 2.8.0. 

All the best, 
Florian

**Comment by jokergoo** [comment, marked as answer]
I think I only tested `HeatmapAnnotation()` with `lapply()` and `do.call()` while not with `mapply()`. I will have a check.

**Reply by flde** [reply]
Many thanks, I think it is always possible to adapt lappyl() or do.cal() so I mark as answered. 

Maybe one comment. If I use as.grob(expression(plot(object))) within a lapply() function, as.grob will not find object. I have to assign the objec to the global environment first within the function call by object <<- object.

**Reply by jokergoo** [reply]
I have never used it like that.

Why don't you use `grid.grabExpr(plot(object))`?

**Comment by jokergoo** [comment]
What is your purpose of doing the following two?

```
lapply(list(a=1), plot_FUN)
```

and 

```
plot_FUN <- function(x, y) {HeatmapAnnotation(foo1 = runif(10))}
mapply(plot_FUN, list(a=1), list(a=1), SIMPLIFY = FALSE)
```

Do you want to automatically generate a list of annotations? Why don't you use `do.call()`?

```
do.call("HeatmapAnnotation", list(a = 1:10, b = letters[1:10]))
```

**Comment by pormr** [comment]
I've encountered the same problem recently. After some sort of investigations, I found out that `purrr::map2` can be used as a drop-in replacement of `mapply` in some cases. For more than two arguments, you can also use `foreach` function from `foreach` package.

**Reply by jokergoo** [reply]
I think the problem is I want to automatically identify what users have specified in `HeatmapAnnotation()` function. I can correctly capture everything when `HeatmapAnntation()` is directly used, but I haven't considered complex use cases like with `mapply()` or maybe `sapply()` also.

**Reply by jokergoo** [reply]
These scenarios I have already tested:

```r
HeatmapAnnotation(a = 1:10)
rowAnnotation(a = 1:10)
columnAnnotation(a = 1:10)
do.call(HeatmapAnnotation, list(a = 1:10))
do.call(rowAnnotation, list(a = 1:10))
do.call(columnAnnotation, list(a = 1:10))
do.call("HeatmapAnnotation", list(a = 1:10))
do.call("rowAnnotation", list(a = 1:10))
do.call("columnAnnotation", list(a = 1:10))
```

But I haven't tested these scenarios:

```r
f = function() HeatmapAnnotation(a = 1:10)
f()
f = function() rowAnnotation(a = 1:10)
f()
f = function() columnAnnotation(a = 1:10)
f()

sapply(1, function(x) HeatmapAnnotation(a = 1:10))
sapply(1, function(x) rowAnnotation(a = 1:10))
sapply(1, function(x) columnAnnotation(a = 1:10))

mapply(function(x) HeatmapAnnotation(a = 1:10), list(1), list(1))
mapply(function(x) rowAnnotation(a = 1:10), list(1), list(1))
mapply(function(x) columnAnnotation(a = 1:10), list(1), list(1))
```

I will have a close look.

**Reply by jokergoo** [reply]
@flde @pormr In the previous versions, I made validations on annotations too complicated. Now I simplified them and it is working now. PS: I still need to make more tests to ensure there is no more error, but at least your example can be run successfully.

```
> plot_FUN <- function(x, y) {HeatmapAnnotation(foo1 = runif(10))}
> mapply(plot_FUN, list(a=1), list(a=1), SIMPLIFY = FALSE)
$a
A HeatmapAnnotation object with 1 annotation
  name: heatmap_annotation_2
  position: column
  items: 10
  width: 1npc
  height: 5mm
  this object is subsettable
  9.11106666666666mm extension on the right

 name   annotation_type color_mapping height
 foo1 continuous vector        random    5mm
```
